scalableminds/webknossos
# Fix buckets loading in the wrong order

When loading a dataset, a portion of the YZ and XZ viewports has always loaded later than it was supposed to be - indicated by large gray stripes or data in a bad resolution. In theory, both viewports should also load data from the middle to the sides, however, this only seemed to work properly for the XY viewport. See the issue in this gif:

![before](https://user-images.githubusercontent.com/1702075/139315176-c973be36-1f41-4e20-94a3-8442db7e4138.gif)

This has been annoying me for ages and I finally decided to get to the bottom of it.
After quite an extensive debugging session, I found the culprit. When picking buckets, only the first occurrence of each bucket (and its priority) was saved to a queue and later added to the push queue. In conjunction with the feature that preloads data from the next slice with lower priority, this led to some buckets in the YZ and XZ viewports always being loaded with lower priority. Although they were encountered later with higher priority, this was not propagated.

Since items in a priority queue cannot be altered or easily removed, I fixed that by keeping track of the highest (actually lowest) priority of each encountered bucket in the bucket picking code and then only adding buckets to the priority queue in the very end.

Result:

![after](https://user-images.githubusercontent.com/1702075/139319549-e7df95d6-00af-474f-b530-6ee83333af37.gif)

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Open a dataset with one mag and reload the color layer data. Data should load from center to the sides in all three viewports.

------
